### ISSUE TITLE
npm ERR! Error: EACCES: permission denied when using Yo Hubot

### ISSUE THREAD

**Issue by DeanMW** [yeoman/yeoman.github.io#619]
Hi,

I have looked at many many other tutorials on setting this up and fixing these errors. I have managed  by following: 

https://github.com/krakenjs/generator-kraken/issues/114#issuecomment-54201366
https://github.com/yeoman/yo/issues/348
https://github.com/keystonejs/keystone/issues/1566

These have fixed the config store access denied errors but now I am getting npm access errors shown below.

root@deanmw-debian:/home/Documents/Programs/hubot# sudo yo hubot
? ==========================================================================
We're constantly looking for ways to make yo better! 
May we anonymously report usage statistics to improve the tool over time? 
More info: https://github.com/yeoman/insight & http://yeoman.io
========================================================================== No
                     _____________________________  
                    /                             \ 
   //\              |      Extracting input for    |
  ////\    _____    |   self-replication process   |
 //////\  /_____\   \                             / 
 ======= |[^_/\_]|   /----------------------------  
  |   | _|___@@__|__                                
  +===+/  ///     \_\                               
   | |_\ /// HUBOT/\\                             
   |___/\//      /  \\                            
         \      /   +---+                            
          \____/    |   |                            
           | //|    +===+                            
            \//      |xx|                            

? Owner deanmwake@email.com
? Bot name this-setup-sucks
? Description A simple helpful robot for your Company
? Bot adapter campfire
   create bin/hubot
   create bin/hubot.cmd
   create Procfile
   create README.md
   create external-scripts.json
   create hubot-scripts.json
identical .gitignore
   create package.json
   create scripts/example.coffee
identical .editorconfig
                     _____________________________  
 _____              /                             \ 
 \    \             |   Self-replication process   |
 |    |    _____    |          complete...         |
 |__\\|   /_____\   \     Good luck with that.    / 
   |//+  |[^_/\_]|   /----------------------------  
  |   | _|___@@__|__                                
  +===+/  ///     \_\                               
   | |_\ /// HUBOT/\\                             
   |___/\//      /  \\                            
         \      /   +---+                            
          \____/    |   |                            
           | //|    +===+                            
            \//      |xx|                            

npm ERR! Linux 3.16.0-4-686-pae
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install" "hubot" "hubot-scripts" "hubot-diagnostics" "hubot-help" "hubot-heroku-keepalive" "hubot-google-images" "hubot-google-translate" "hubot-pugme" "hubot-maps" "hubot-redis-brain" "hubot-rules" "hubot-shipit" "--save"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! path /root/.npm/redis/0.8.4/package.tgz.3721715593
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall open

npm ERR! Error: EACCES: permission denied, open '/root/.npm/redis/0.8.4/package.tgz.3721715593'
npm ERR!     at Error (native)
npm ERR!  { [Error: EACCES: permission denied, open '/root/.npm/redis/0.8.4/package.tgz.3721715593']
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/root/.npm/redis/0.8.4/package.tgz.3721715593',
npm ERR!   parent: 'hubot-redis-brain' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/Documents/Programs/hubot/npm-debug.log

**Comment by SBoudrias** [comment]
That's your error `npm ERR! Error: EACCES: permission denied, open '/root/.npm/redis/0.8.4/package.tgz.3721715593'`

You should google chmod.

This bug is related either to your computer or to npm. If you need further support use their channels.

**Comment by DeanMW** [comment]
I have followed: https://github.com/yeoman/yeoman.github.io/issues/282 and many other tutorials and am still getting NPM issues. Why not just let yo be run as root. This would solve almost all of these countless issues people are having.

**Comment by SBoudrias** [comment]
@DeanMW no it would not. I would bring back the biggest problem we had before of user unable to edit their files.

**Comment by DeanMW** [comment]
after creating a new user, other than root and running the command I am still getting:

npm ERR! Linux 3.16.0-4-686-pae
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install" "hubot" "hubot-scripts" "hubot-diagnostics" "hubot-help" "hubot-heroku-keepalive" "hubot-google-images" "hubot-google-translate" "hubot-pugme" "hubot-maps" "hubot-redis-brain" "hubot-rules" "hubot-shipit" "--save"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! path /home/Documents/Programs/hubot
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall access

npm ERR! Error: EACCES: permission denied, access '/home/Documents/Programs/hubot'
npm ERR!     at Error (native)
npm ERR!  { [Error: EACCES: permission denied, access '/home/Documents/Programs/hubot']
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/home/Documents/Programs/hubot' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.
npm ERR! Linux 3.16.0-4-686-pae
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install" "hubot" "hubot-scripts" "hubot-diagnostics" "hubot-help" "hubot-heroku-keepalive" "hubot-google-images" "hubot-google-translate" "hubot-pugme" "hubot-maps" "hubot-redis-brain" "hubot-rules" "hubot-shipit" "--save"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! path npm-debug.log.3490650549
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall open

npm ERR! Error: EACCES: permission denied, open 'npm-debug.log.3490650549'
npm ERR!     at Error (native)
npm ERR!  { [Error: EACCES: permission denied, open 'npm-debug.log.3490650549']
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: 'npm-debug.log.3490650549' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/Documents/Programs/hubot/npm-debug.log

**Comment by TinajaLabs** [comment]
Raspberry Pi - came with node v10.29 but not the right version for yeoman, went through a lot of issues to get a newer version.

Now I go to install hubot - https://hubot.github.com/docs/

Instructions are:

You will need node.js and npm. Once those are installed, we can install the hubot generator:
```
%  npm install -g yo generator-hubot
```
This will give us the hubot yeoman generator. Now we
can make a new directory, and generate a new instance of hubot in it. For example, if
we wanted to make a bot called myhubot:
```
% mkdir myhubot
% cd myhubot
% yo hubot
```

Nothing about sudo.

After installing yo and generator-hubot, I get this:
```
root@bot:~# npm install -g yo generator-hubot                                                               
npm WARN deprecated CSSselect@0.4.1: the module is now available as 'css-select'
npm WARN deprecated CSSwhat@0.4.7: the module is now available as 'css-what'
npm WARN deprecated npmconf@2.1.2: this package has been reintegrated into npm and is now out of date with respect to npm
/usr/local/bin/yo -> /usr/local/lib/node_modules/yo/lib/cli.js

> yo@1.7.0 postinstall /usr/local/lib/node_modules/yo
> yodoctor


Yeoman Doctor
Running sanity checks on your system

✔ Global configuration file is valid
✔ Node.js version
✔ No .bowerrc file in home directory
✔ No .yo-rc.json file in home directory
✔ npm version
✔ NODE_PATH matches the npm root

Everything looks all right!                                                                                                                                             
/usr/local/lib
├── generator-hubot@0.3.1 
└── yo@1.7.0 

```

It was a scramble to get "stable" version of node using 'n'.  Was I supposed to run all that as user pi?  Info indicates I'm supposed to install node as root.

I often run as root.  Seem to have problems when trying to use some other user.  If I create the files as root, won't I be able to edit as root?

What I'd like is a tutorial on how I should do what you recommend; run as pi 
to uninstall node that comes with raspberry pi, 
to install node 0.12.x, 
to install yeoman,
to install hubot; 
and when to use user, pi and when to use sudo.

I got Rocket.chat and mongodb installed on the raspberry pi and want to hook-in hubot.  Most of my energy is being used to get yeoman and the hubot generator configured.

Sadly I get a series of errors like:
```
Error: EACCES: permission denied, open '/opt/myhubot/bin/hubot'
```
I create the required folders and re-run the generator, only to be stumped with the next permission denied.  And this as root.  How can I be denied if I'm running as root?

Sorry for the frustration...

**Comment by SBoudrias** [comment]
These are all issues with npm. Try to get support from their community.

**Comment by xMTinkerer** [comment]
@TinajaLabs, I fixed this by removing npm and all the global modules per [this](http://stackoverflow.com/questions/9283472/command-to-remove-all-npm-modules-globally/) post. 
Then I installed [nvm](https://github.com/creationix/nvm/blob/master/README.markdown). Then running `yo hubot` did not generate the permission denied error.

**Comment by TinajaLabs** [comment]
Thanks, xMTinkerer,
I'll try that next time I install Hubot.  I muddled through it all and eventually got my app to work.

For my Home Automation project, I installed Rocket.chat and Hubot to communicate with my coordinating piece, Node-Red, and was able to chat with my home. 

From a command typed into Rocket.Chat --> Hubot --> MQTT --> Node-Red --> MQTT --> lights on/off (configured X10, and ESP8266 devices) and confirming message returned.  For me, it is a thing of beauty.  ;)
